Produce light particles from decay

paulromano opened the pull request:
A user [recently noted](https://openmc.discourse.group/t/he-4-production-from-alpha-decay-and-n-alpha-reactions/2473) that decay calculations do not account for the production of light particles. For example, if you were to look at the depletion chain data for <sup>5</sup>Li:
```xml
<nuclide name="Li5" half_life="3.06868e-22" decay_modes="1" decay_energy="1965000.0" reactions="0">
  <decay type="alpha" target="H1" branching_ratio="1.0"/>
</nuclide>
```
running a decay calculation would produce H1 because it's explicitly listed as a "target", but the alpha particle itself would not be produced. This PR fixes this so that alphas and/or protons are produced when the decay mode indicates it.

I've also made some changes to allow `IndependentOperator` to work with an empty `MicroXS` object (useful for doing a decay-only calculation). @yardasol would you be able to review the changes here?

yardasol commented on tests/unit_tests/test_deplete_decay_products.py:
```diff
@@ -0,0 +1,43 @@
+        normalization_mode='source-rate'
+    )
+
+    # Create time-integrator and integrate
+    integrator = openmc.deplete.PredictorIntegrator(
+        op, timesteps=[1.0], source_rates=[0.0], timestep_units='d'
+    )
+    integrator.integrate(final_step=False)
+
+    # Get concentration of H1 and He4
+    results = openmc.deplete.Results('depletion_results.h5')
+    _, h1 = results.get_atoms("1", "H1")
+    _, he4 = results.get_atoms("1", "He4")
+
+    # Since we started with 1e24 atoms of Li5, we should have 1e24 atoms of both
+    # H1 and He4
+    assert h1[1] == pytest.approx(1e24)
+    assert he4[1] == pytest.approx(1e24)
```
I like this a lot.

yardasol reviewed (changes requested):
Hi @paulromano,

Great work here. I know in the past we have found that our test suite is missing some cases. In your changes to the `StepResult` class, have you ensured that there are sufficient test cases to check the changes you have added? If not, we should probably add those.

My other comments are below. It's been a while since I've worked with the `TransportOperator` classes so I'm a bit rusty on the program flow through them (even though I refactored them into what they are now, ha!), so I made some incorrect statements, but I'm leaving them anyways with some extra clarifying comments.

paulromano commented on openmc/deplete/stepresult.py:
```diff
@@ -318,10 +318,11 @@ def _write_hdf5_metadata(self, handle):
-                              maxshape=(None, n_stages, n_mats, n_nuc_rxn, n_rxn),
-                              chunks=(1, 1, n_mats, n_nuc_rxn, n_rxn),
-                              dtype='float64')
+        if n_nuc_rxn > 0 and n_rxn > 0:
+            handle.create_dataset("reaction rates", (1, n_stages, n_mats, n_nuc_rxn, n_rxn),
+                                maxshape=(None, n_stages, n_mats, n_nuc_rxn, n_rxn),
+                                chunks=(1, 1, n_mats, n_nuc_rxn, n_rxn),
+                                dtype='float64')
```
It's a little stronger than just not writing an empty dataset -- if no reactions exist on the `MicroXS` object, this will actually fail because `h5py` will refuse to create a dataset where one of the dimensions has size 0.

paulromano commented:
@yardasol I've updated the docstrings to better explain what happens when `chain_file` is not specified and how an empty `MicroXS` object can be used for a decay-only calculation. I believe I've addressed all comments but let me know if anything else needs updating.

yardasol commented on openmc/deplete/independent_operator.py:
```diff
@@ -31,16 +31,21 @@ class IndependentOperator(OpenMCOperator):
     passed to an integrator class, such as
     :class:`openmc.deplete.CECMIntegrator`.
 
+    Note that passing an empty :class:`~openmc.deplete.MicroXS` instance to the
+    ``micro_xs`` argument allows a decay-only calculation to be run.
```
Good

yardasol commented on openmc/deplete/independent_operator.py:
```diff
@@ -31,16 +31,21 @@ class IndependentOperator(OpenMCOperator):
     micro_xs : MicroXS
-        One-group microscopic cross sections in [b] .
+        One-group microscopic cross sections in [b]. If the
+        :class:`~openmc.deplete.MicroXS` is empty, a decay-only calculation will
```
```suggestion
        :class:`~openmc.deplete.MicroXS` object is empty, a decay-only calculation will
```

yardasol commented on openmc/deplete/independent_operator.py:
```diff
@@ -169,10 +174,13 @@ def from_nuclides(cls, volume, nuclides,
-            Path to the depletion chain XML file.
-        nuc_units : {'atom/cm3', 'atom/b-cm'}
+            One-group microscopic cross sections in [b]. If the
+            :class:`~openmc.deplete.MicroXS` is empty, a decay-only calculation
```
```suggestion
            :class:`~openmc.deplete.MicroXS` object is empty, a decay-only calculation
```

yardasol reviewed (approved):
Good docstring additions. I just have a couple tiny suggestions, but they aren't critical so I'm going to approve.

yardasol commented:
Nice work @paulromano
